File: 678.cpp
class Solution {
public:
    bool checkValidString(string s) {
        int countAstrik = 0, countLeft = 0;
        vector<int> stackAstrik, stackleft;
        for (int i = 0; i < s.length(); i++) {
            char ch = s[i];
            if ((countAstrik == 0 && countLeft == 0) && ch == ')') {
                return false;
            } else if (ch == ')') {
                if (countLeft > 0) {
                    countLeft--;
                    stackleft.pop_back();
                } else {
                    countAstrik--;
                    stackAstrik.pop_back();
                }
            } else {
                if (ch == '(') {
                    ++countLeft;
                    stackleft.push_back(i);
                } else {
                    ++countAstrik;
                    stackAstrik.push_back(i);
                }
            }
        }
        while (!stackleft.empty()) {
            if (stackAstrik.empty()) {
                return false;
            } else if (stackleft.back() < stackAstrik.back()) {
                stackleft.pop_back();
                stackAstrik.pop_back();
            } else {
                return false;
            }
        }
        return true;        
    }
};

File: 678.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        countAstrik , countLeft = 0, 0
        stackAstrik = []
        stackleft = []
        for i in range(len(s)):
            if (countAstrik ==0 and countLeft == 0) and s[i]==")":
                return False
            elif s[i] == ")":
                if countLeft >0:
                    countLeft-=1
                    stackleft.pop()
                else:
                    countAstrik-=1
                    stackAstrik.pop()
            else:
                if s[i] =="(":
                    countLeft+=1
                    stackleft.append(i)
                else:
                    countAstrik +=1
                    stackAstrik.append(i)
        while stackleft:
            if len(stackAstrik) == 0:
                return False
            elif stackleft[-1] < stackAstrik[-1]:
                stackleft.pop()
                stackAstrik.pop()
            else:
                return False
 
        return True